File: nodejs/src/crawler.ts
import { extractHrefs } from './crawler_utils.js';
import crawler_page from './crawler_page.js';
import crawler_result from './crawler_result.js';
import puppeteer, { Browser } from 'puppeteer';
import sharp from 'sharp';
import util from 'util';
import path from 'path';
import fs from 'fs';
import crawler_settings from './crawler_settings.js';

export default class crawler {
    private browser: Browser
    private settings: crawler_settings
    private result: crawler_result

    private async initialize(settings: crawler_settings) {
        this.settings = settings;
        this.browser = await puppeteer.launch({
            headless: this.settings.isHeadlessSet(),
            defaultViewport: { width: 0, height: 0},
            args: []
        });
        this.result = new crawler_result(this.settings.getSeedUrl());
    }

    private async uninitialize() {
        await this.browser.close();
    }

    private async crawlPage(page: crawler_page, depth = 0)
    {
        if(depth > this.settings.getMaxDepth()) {
            return;
        }

        if(this.result.hasCrawledPage(page.getUrl())) {
            console.log(`Reusing route: ${page.getUrl()}`);

            const item = this.result.getCrawledPage(page.getUrl());
            page.setTitle(item.getTitle());
            page.setImg(item.getImg());
            page.setImgFilePath(item.getImgFilePath());
            page.setHtml(item.getHtml());
            page.setHtmlFilePath(item.getHtmlFilePath());
            page.setChildren(item.getChildren());

            // Fill in the children with details (if they already exist).
            debugger;
            page.getChildren().forEach(c => {
              const item = this.result.getCrawledPage(c.getUrl());
              if(item !== undefined) {
                c.setTitle(item.getTitle());
                c.setImg(item.getImg());
              }
            });

            return;
        } else {
            console.log(`Loading: ${page.getUrl()}`);

            const newPage = await this.browser.newPage();
            await newPage.goto(page.getUrl(), { waitUntil: 'networkidle2' });

            let hrefs = await extractHrefs(newPage) as Array<string>;
            hrefs = hrefs.filter((url: string) => {
                // Filter out urls that are the same url as seed.
                if((url.replace(/\/+$/, '') !== this.settings.getSeedUrl().replace(/\/+$/, '')))
                {
                    return true;
                }
                // Filer out urls that do not fit the crawl regex.
            });

            page.setUrl(await newPage.url());
            page.setTitle(await newPage.evaluate('document.title') as string);

            hrefs.forEach(link => {
                page.addChild(new crawler_page(link));                
            });

            // screenshots
            if(this.settings.isScreenShotsSet()) {
                var imgFilepath = path.join(this.result.getOutputFolder(), `${page.getGuid()}_screenshot.png`);
                let imgBuffer = await sharp(await newPage.screenshot()).resize(null, 600).toBuffer();
                util.promisify(fs.writeFile)(imgFilepath, imgBuffer);
                page.setImg(`data:img/png;base64,${imgBuffer.toString('base64')}`);
                page.setImgFilePath(imgFilepath);
            }

            // dom dumps
            if(this.settings.isDumpDomSet()) {
                var htmlFilepath = path.join(this.result.getOutputFolder(), `${page.getGuid()}_dom.html`);
                let html = await newPage.content();
                util.promisify(fs.writeFile)(htmlFilepath, html);
                page.setHtml(html);
                page.setHtmlFilePath(htmlFilepath);
            }

            // other things we may want to do on each page visited

            this.result.addCrawledPage(page.getUrl(), page);
            await newPage.close();
        }

        for(const child of page.getChildren()) {
            await this.crawlPage(child, depth + 1);
        }
    }

    async run(settings: crawler_settings) {
        await this.initialize(settings);
        await this.crawlPage(new crawler_page(this.settings.getSeedUrl()));
        await this.result.writeResults();
        await this.uninitialize();
    }
}

File: dotnet/Crawler.cs
using PuppeteerSharp;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SpaCrawler
{
    public class Crawler
    {
        private const string _collectAllAnchorsDeepScriptJs =
            @"(sameOrigin = true) => {
                const allElements = [];

                const findAllElements = function(nodes) {
                  for (let i = 0, el; el = nodes[i]; ++i) {
                    allElements.push(el);
                    if (el.shadowRoot) {
                      findAllElements(el.shadowRoot.querySelectorAll('*'));
                    }
                  }
                };

                findAllElements(document.querySelectorAll('*'));

                const filtered = allElements
                  .filter(el => el.localName === 'a' && el.href)
                  .filter(el => el.href !== location.href)
                  .filter(el => {
                    if (sameOrigin) {
                      return new URL(location).origin === new URL(el.href).origin;
                    }
                    return true;
                  })
                  .map(a => a.href);

                return Array.from(new Set(filtered));
            }
            ";
        private Browser browser;
        private CrawlSettings settings;
        private CrawlResult result;

        private async Task InitializeCrawl(CrawlSettings settings)
        {
            this.settings = settings;
            await new BrowserFetcher().DownloadAsync(BrowserFetcher.DefaultRevision);
            browser = await Puppeteer.LaunchAsync(new LaunchOptions
            {
                Headless = settings.Headless,
                DefaultViewport = new ViewPortOptions { Width = 0, Height = 0 },
                Args = new string[] { "" }
            });
            result = new CrawlResult(settings.SeedUrl.ToString());
        }
        private async Task UninitializeCrawl()
        {
            await browser.CloseAsync();
        }
        private async Task CrawlPageAsync(CrawlPage pageToCrawl, uint depth = 0)
        {
            if (depth > settings.MaxDepth)
            {
                return;
            }

            CrawlPage crawledPage;
            if(result.CrawledPages.TryGetValue(pageToCrawl.Url.ToString(), out crawledPage))
            {
                Console.WriteLine(string.Format("Resuing route: {0}", pageToCrawl.Url));

                pageToCrawl.Title = crawledPage.Title;
                pageToCrawl.Img = crawledPage.Img;
                pageToCrawl.Dom = crawledPage.Dom;
                pageToCrawl.ChildPages = crawledPage.ChildPages;

                foreach(var childPage in pageToCrawl.ChildPages)
                {
                    if(result.CrawledPages.TryGetValue(childPage.Url.ToString(), out crawledPage))
                    {
                        childPage.Title = crawledPage.Title;
                        childPage.Img = crawledPage.Img;
                    }
                }
                return;
            }
            else
            {
                Console.WriteLine(string.Format("Loading page: {0}", pageToCrawl.Url));

                Page newPage = await browser.NewPageAsync();
                await newPage.GoToAsync(pageToCrawl.Url.ToString(), WaitUntilNavigation.Networkidle2);

                var anchorsJsToken = await newPage.EvaluateFunctionAsync(_collectAllAnchorsDeepScriptJs, true);
                var hrefs = anchorsJsToken.ToObject<HashSet<string>>();
                hrefs = hrefs.Where(a => a != settings.SeedUrl.ToString()).ToHashSet();

                var title = await newPage.EvaluateExpressionAsync("document.title");
                pageToCrawl.Title = title.ToString();

                foreach(var href in hrefs)
                {
                    pageToCrawl.ChildPages.Add(new CrawlPage(new Uri(href)));
                }

                if(settings.ScreenShots)
                {
                    string imageFile = string.Format(@"{0}_screenshot.png", pageToCrawl.Guid.ToString().Replace("-", ""));
                    Bitmap bitMap = new Bitmap(Image.FromStream(await newPage.ScreenshotStreamAsync()));
                    imageFile = Path.Join(result.OutputFolder.FullName, imageFile);
                    bitMap.Save(imageFile);
                    pageToCrawl.Img = new CrawlImg(bitMap, imageFile);
                }

                if(settings.DumpDom)
                {
                    string domFile = string.Format(@"{0}_dom.html", pageToCrawl.Guid.ToString().Replace("-", ""));
                    var domHtml = await newPage.GetContentAsync();
                    domFile = Path.Join(result.OutputFolder.FullName, domFile);
                    File.WriteAllText(domFile, domHtml);
                    pageToCrawl.Dom = new CrawlDom(domHtml, domFile);
                }

                result.CrawledPages.Add(pageToCrawl.Url.ToString(), pageToCrawl);
                await newPage.CloseAsync();
            }

            foreach(var childPage in pageToCrawl.ChildPages)
            {
                await CrawlPageAsync(childPage, depth + 1);
            }
        }
        public async Task RunAsync(CrawlSettings settings)
        {
            await InitializeCrawl(settings);
            await CrawlPageAsync( new CrawlPage(new Uri(settings.SeedUrl.ToString())));
            result.WriteResults();
            await UninitializeCrawl();
        }
    }
}
